AI/VCAI/VCAI.cpp

```diff
@@ -1844,7 +1844,7 @@ bool VCAI::isAccessibleForHero(const int3 & pos, HeroPtr h, bool includeAllies)
 			}
 		}
 	}
-	return cb->getPathsInfo(h.get())->getPathInfo(pos)->reachable();
+	return ah->isTileAccessible(h, pos);

```

Comment:
This is not equivalent. PathfindingManager will return true if tile is accessible through a battle or any different way and reachable() returns true only if it is directly accessible. But in general it should be replaced.

Reply:
Is there an equivalent in AIPathfinder?

Reply:
Btw VisitTile itself is an example of code which will defenitelly break if swithced to 
ah->isTileAccessibleForHero because it expects direct access

Reply:
Technically it is possible to implement equivalent function. AIPathfinder nodes are tiles where a hero has to do something like battle or spell cast. So they are not neighbors like in normal CPathfinder. In order to move between them we still need to call cb->getPathsInfo() but only at tryRealize(VisitTile) when we actually decided to move a hero.

So if AIPathfinder path has the length of 1 it is directly accessible. We can implement another version of isAccessibleForHero and check any path length to be 1. 

But this is not needed. We just need to avoid using VisitTile/VisitObj directly but use ah->howToVisitTile/Obj instead. If it returns an empty vector of goals the object is not accessible. So we need to replace all direct usages of VistTile/Obj with ah->howToVisitTile/Obj whenever it is used and remove all logic from VisitTile regarding to checking if tile is accessible and invoking ClearWayTo. But this is my opinion and it is a lot of work.

Reply:
The actual problem is lack of paths cache in Client, single hero cache misses a lot (10 times more path recalculations that should be).

Reply:
Or you can do it yourself. The same way as it is already done but you need to check that any node also has no previousNeighbor and specialAction

Reply:
I will do something anyway, still not decided what exactly, may be even both.